### DISCUSSION TITLE
How to cleanly use UUIDs in User model without everything breaking?

### DISCUSSION THREAD

**Discussion by caposaric** [orchidsoftware/platform#2532] [Q&A]
I'm trying to make my Dashboard play along with Orchid's current implementation, but having my `User` model use uuid instead of integer as primary key. I managed to do it using Laravel 9's `HasUuids` trait and replacing the model class as specified [here](https://orchid.software/en/docs/configuration/#model-classes). This part seems to be fine.

Unfortunately, it seems like this change does not play well with what's already there, as I'm stumbling with error after error, namely:
- Relation to `Role` model seems to break - probably because `role_users` table has a composite key (`['user_id', 'role_id']`). I tried casting `user_id` only to string, making an intermediate (extending `Pivot`) table to cast the same field and several other methods... Can't get it working. Decided to leave Roles alone (will need them in the future but not strictly necessary right now) and move on, but...
- Trying to create a route with this User model as a parameter is not working, as the `$user->id` (type _uuid_) property is being casted to an integer and trimmed at the first non-numeric character. Also, if I manually replace the ID on the URL and get to the `UserEditScreen` I go back to problem number 1 because I can't load Roles.

Just these 2 things make it practically impossible to use the existing User/Role CRUD boilerplate after that small change. Are there any suggestions in these regard? Am I missing something obvious here?

**Comment by caposaric** [comment]
Just following up on my own discussion..
- The `Dashboard` import line on the model class replacement example was wrong and causing my `App\Models\User` not to be picked up, hence its `$casts` not working at all. Using the right import fixed the issue I had with `route()` helper casting the uuid to an integer. (Opened a PR to update this)
- Replaced all imports of Orchid's User class on `UserXYZScreen`s to use `App\Models\User` and it seems to be working just fine.

Hopefully this solves all issues I was having. Feedback welcome. :)

**Comment by tabuna** [comment]
- To solve the use of UUID, create a migration with something like this:

```php
/**
 * Run the migrations.
 *
 * @return void
 */
public function up()
{
    Schema::table('role_users', function (Blueprint $table) {
        $table->char('user_id')->change();
    });
}
```


- Don't forget to put your custom model in the same way:

```php
/**
 * Get the value indicating whether the IDs are incrementing.
 *
 * @return bool
 */
public function getIncrementing()
{
    return false;
}

/**
 * Get the auto-incrementing key type.
 *
 * @return string
 */
public function getKeyType()
{
    return 'string';
}
```

- Make sure that your model is used in the existing files and not the pre-installed one (`use App/Models/User`, not `use Orchid\Platform\Models\User`)
